src/core/public/http/types.ts

```diff
@@ -298,9 +325,9 @@ export interface HttpInterceptor {
    * @param controller {@link IHttpInterceptController}
    */
   request?(
-    request: Request,
+    fetchOptions: Readonly<HttpFetchOptionsWithPath>,
     controller: IHttpInterceptController
-  ): Promise<Request> | Request | void;
+  ): Promise<Partial<HttpFetchOptionsWithPath>> | Partial<HttpFetchOptionsWithPath> | void;
```

Comment:
NIT: for readability, maybe we want to introduce `type MaybePromise<T> = Promise<T> | T;` ?

Reply:
Added one to `@kbn/utility-types`.